metriq_gym/quantinuum/device.py

```diff
@@ -0,0 +1,92 @@
+from typing import Any
+import os
+import uuid
+from datetime import datetime, timezone
+
+import qnexus as qnx
+from qnexus.models.language import Language
+from pytket import Circuit
+from pytket.extensions.qiskit import qiskit_to_tk
+from qiskit import QuantumCircuit
+
+from qbraid import QPROGRAM
+from qbraid.runtime import DeviceStatus, QuantumDevice, TargetProfile
+from qbraid.programs import ExperimentType, ProgramSpec
+from metriq_gym.quantinuum.job import QuantinuumJob
+
+
+def _profile(device_id: str) -> TargetProfile:
+    return TargetProfile(
+        device_id=device_id,
+        # TODO: find a property to distinguish real vs. simulator
+        simulator="E" in device_id.upper(),
+        experiment_type=ExperimentType.GATE_MODEL,
+        num_qubits=None,
+        program_spec=ProgramSpec(Circuit),
+        basis_gates=None,
+        provider_name="quantinuum",
+        extra={},
+    )
+
+
+class QuantinuumDevice(QuantumDevice):
+    def __init__(self, *, provider, device_id: str) -> None:
+        super().__init__(_profile(device_id))
+        self._provider = provider
+
+    def status(self) -> DeviceStatus:
+        return DeviceStatus.ONLINE
+
+    def transform(self, run_input: Any):
+        if isinstance(run_input, Circuit):
+            return run_input
+        if isinstance(run_input, list):
+            return [self.transform(item) for item in run_input]
+        if isinstance(run_input, QuantumCircuit):
+            return qiskit_to_tk(run_input)
+        raise TypeError(
+            f"Unsupported run_input type {type(run_input)}; expected pytket.Circuit or qiskit.QuantumCircuit"
+        )
+
+    def submit(self, run_input: QPROGRAM, *, shots: int | None = None, **_: Any) -> QuantinuumJob:
+        project = qnx.projects.get_or_create(name=os.getenv("QNEXUS_PROJECT_NAME", "metriq-gym"))
+        backend_config = qnx.QuantinuumConfig(device_name=self.profile.device_id)
+
+        circuits = self.transform(run_input)
+        circuits_list = circuits if isinstance(circuits, list) else [circuits]
+
+        def unique(label: str) -> str:
+            ts = datetime.now(timezone.utc).strftime("%Y%m%dT%H%M%S")
+            return f"metriq-gym {label} {ts}-{uuid.uuid4().hex[:6]}"
+
+        circuit_refs = [
+            qnx.circuits.upload(name=unique(f"circuit-{i}"), circuit=c, project=project)
+            for i, c in enumerate(circuits_list)
+        ]
+
+        opt = int(os.getenv("QNEXUS_OPT_LEVEL", "1"))
+        compile_job = qnx.start_compile_job(
+            programs=circuit_refs,
+            name=unique("compile"),
+            optimisation_level=opt,
+            backend_config=backend_config,
+            project=project,
+        )
+        qnx.jobs.wait_for(compile_job)
```

Comment:
We should leave a note here (and maybe even a log statement) that this is blocking. Since this will be called in the dispatch actions, it may be confusing for the user in case it hangs.